backend/src/ml_space_lambda/utils/iam_manager.py

```diff
@@ -435,21 +444,84 @@ def _detach_iam_policies(self, iam_role_name: str) -> List[str]:
 
         return detached_iam_policies
 
-    def _generate_project_policy(self, project: str) -> str:
-        return (
-            self.project_policy.replace("\n", "")
-            .replace("$PROJECT_NAME", project)
-            .replace("$BUCKET_NAME", self.data_bucket)
-            .replace("$PARTITION", self.aws_partition)
-        )
+    """
+    Generates a policy from the provided template and variable replacments
 
-    def _generate_user_policy(self, user: str) -> str:
-        return (
-            self.user_policy.replace("\n", "")
-            .replace("$USER_NAME", user)
-            .replace("$BUCKET_NAME", self.data_bucket)
-            .replace("$PARTITION", self.aws_partition)
-        )
+    Inputs:
+    policy_template: str - A template that includes variables of the form $VAR_NAME
+    replacments: list[list[str]] - A list of str tuples that are to be replaced of the form ["$VAR_NAME", "var value"]
+    """
+
+    def _generate_policy_from_template_string(self, policy_template: str, replacements: list[list[str]]):
+        policy = policy_template.replace("\n", "")
+        for replacement in replacements:
+            policy = policy.replace(replacement[0], replacement[1])
+        return policy
+
+    def generate_policy(self, statements: list):
+        if len(statements) > 0:
+            return {"Version": "2012-10-17", "Statement": statements}
+        else:
+            return None
+
+    def generate_policy_string(self, statements: list):
+        policy = self.generate_policy(statements)
+        return json.dumps(policy) if policy is not None else None
+
+    # If the provided policy doesn't exist, it will be created
+    def update_dynamic_policy(
+        self,
+        policy: str,
+        policy_name: str,
+        policy_type: str,
+        policy_type_identifier: str,
+        on_create_attach_to_notebook_role: bool = False,
```

Comment:
Dont we want this to always be true? what would be a situation where we wont want to add it to the notebook role?

Reply:
Currently all dynamic policies we are creating, we probably do want to attach to the notebook role, but this is siding cautious as we have a limit to the number of policies we can attach and I am attempting to be frugal and not accidentally attach policies we don't intend to by default. The idea is that this should be an intentionally opted in feature as accidentally attaching a policy to the notebook role means it will then be attached to all other dynamic roles in the future.